// @wdio/devtools-service/build/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isBrowserSupported = exports.getBrowserMajorVersion = exports.isBrowserVersionLower = exports.isSupportedUrl = exports.sumByKey = exports.setUnsupportedCommand = void 0;
const constants_1 = require("./constants");
const VERSION_PROPS = ['browserVersion', 'browser_version', 'version'];
const SUPPORTED_BROWSERS_AND_MIN_VERSIONS = {
    'chrome': 63,
    'chromium': 63,
    'googlechrome': 63,
    'google chrome': 63,
    'firefox': 86
};
function setUnsupportedCommand(browser) {
    return browser.addCommand('cdp', /* istanbul ignore next */ () => {
        throw new Error(constants_1.UNSUPPORTED_ERROR_MESSAGE);
    });
}
exports.setUnsupportedCommand = setUnsupportedCommand;
/**
 * Create a sum of a specific key from a list of objects
 * @param list list of key/value objects
 * @param key  key of value to be summed up
 */
function sumByKey(list, key) {
    return list.map((data) => data[key]).reduce((acc, val) => acc + val, 0);
}
exports.sumByKey = sumByKey;
/**
 * check if url is supported for tracing
 * @param  {String}  url to check for
 * @return {Boolean}     true if url was opened by user
 */
function isSupportedUrl(url) {
    return constants_1.IGNORED_URLS.filter((ignoredUrl) => url.startsWith(ignoredUrl)).length === 0;
}
exports.isSupportedUrl = isSupportedUrl;
/**
 * check if browser version is lower than `minVersion`
 * @param {object} caps capabilities
 * @param {number} minVersion minimal chrome browser version
 */
function isBrowserVersionLower(caps, minVersion) {
    const versionProp = VERSION_PROPS.find((prop) => caps[prop]);
    const browserVersion = getBrowserMajorVersion(caps[versionProp]);
    return typeof browserVersion === 'number' && browserVersion < minVersion;
}
exports.isBrowserVersionLower = isBrowserVersionLower;
/**
 * get chromedriver major version
 * @param   {string|*}      version chromedriver version like `78.0.3904.11` or just `78`
 * @return  {number|*}              either major version, ex `78`, or whatever value is passed
 */
function getBrowserMajorVersion(version) {
    if (typeof version === 'string') {
        const majorVersion = Number(version.split('.')[0]);
        return isNaN(majorVersion) ? parseInt(version, 10) : majorVersion;
    }
    return version;
}
exports.getBrowserMajorVersion = getBrowserMajorVersion;
/**
 * check if browser is supported based on caps.browserName and caps.version
 * @param {object} caps capabilities
 */
function isBrowserSupported(caps) {
    if (!caps.browserName ||
        !(caps.browserName.toLowerCase() in SUPPORTED_BROWSERS_AND_MIN_VERSIONS) ||
        isBrowserVersionLower(caps, SUPPORTED_BROWSERS_AND_MIN_VERSIONS[caps.browserName.toLowerCase()])) {
        return false;
    }
    return true;
}
exports.isBrowserSupported = isBrowserSupported;
